packages/sdk/src/types/util.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * Discriminator for {@linkcode TypedJson}. The member used to identify each type is
+ * exported from each type module and can be accessed through named export `JSON_DISCRIMINATOR`.
+ */
+export enum TypedJsonDiscriminator {
+    AccountAddress = 'ccd_account_address',
+    BlockHash = 'ccd_block_hash',
+    CcdAmount = 'ccd_ccd_amount',
+    ContractAddress = 'ccd_contract_address',
+    ContractName = 'ccd_contract_name',
+    CredentialRegistrationId = 'ccd_cred_reg_id',
+    DataBlob = 'ccd_data_blob',
+    Duration = 'ccd_duration',
+    Energy = 'ccd_energy',
+    EntrypointName = 'ccd_entrypoint_name',
+    InitName = 'ccd_init_name',
+    ModuleReference = 'ccd_module_reference',
+    Parameter = 'ccd_parameter',
+    ReceiveName = 'ccd_receive_name',
+    ReturnValue = 'ccd_return_value',
+    SequenceNumber = 'ccd_sequence_number',
+    Timestamp = 'ccd_timestamp',
+    TransactionExpiry = 'ccd_transaction_expiry',
+    TransactionHash = 'ccd_transaction_hash',
+}
+
+/**
+ * Type describing the JSON representation of strong types used in the SDK.
+ *
+ * @template V - The serializable JSON value
+ */
+export interface TypedJson<V> {
+    /** The type discriminator */
+    ['@type']: TypedJsonDiscriminator;
+    /** The serializable type value */
+    value: V;
+}
+
+/**
+ * Base class for concordium domain types
+ *
+ * @template D - The JSON discriminator
+ * @template V - The serializable JSON value
+ */
+export abstract class TypeBase<V> implements ToTypedJson<V> {
+    protected abstract typedJsonType: TypedJsonDiscriminator;
+    protected abstract get serializable(): V;
+
+    public toTypedJSON(): TypedJson<V> {
```

Comment:
Since we are not using the built-in `toJSON` this might as well be a separate function, so it can be tree-shaken.
It is not really any different from the other serializations we have (protobuf, schemaJson, bytes and so on)